File: server/entities/function.go
package entities

import "time"

const (
	RunningFunctionStatus = "RUNNING"
	ErrorFunctionStatus   = "ERROR"
)

type Function struct {
	ID                           string
	Name                         string
	SourceCode                   string
	MethodToExecute              string
	Status                       string
	InputConnectorDataSourceID   string
	InputConnectorConfiguration  map[string]interface{}
	OutputConnectorDataSourceID  string
	OutputConnectorConfiguration map[string]interface{}
	CreatedAt                    time.Time
	UpdatedAt                    *time.Time
}

File: client/src/schemas/function.ts
import { z } from "zod";

import { MQTTConfigSchema, RedisConfigSchema } from "./dataSource";

export const FunctionSchema = z.object({
    id: z.string(),
    name: z.string(),
    sourceCode: z.string(),
    methodToExecute: z.string(),
    status: z.enum(["CREATING", "RUNNING", "ERROR"]),
    inputConnectorType: z.enum(["MQTT", "REDIS"]),
    inputConnectorConfiguration: z.union([MQTTConfigSchema, RedisConfigSchema]),
    outputConnectorType: z.enum(["MQTT", "REDIS"]),
    outputConnectorConfiguration: z.union([MQTTConfigSchema, RedisConfigSchema]),
});
